core/src/Kokkos_CopyViews.hpp

```diff
@@ -2645,11 +2723,58 @@ inline void deep_copy(
         "(none)", &value, dst.span() * sizeof(typename dst_traits::value_type));
   }
 #endif
-  ExecSpace().fence();
-  typedef
-      typename View<DT, DP...>::uniform_runtime_nomemspace_type ViewTypeUniform;
-  Kokkos::Impl::ViewFill<ViewTypeUniform>(dst, value);
-  ExecSpace().fence();
+  if (dst.data() == nullptr) {
+    space.fence();
+  } else {
+    typedef typename View<DT, DP...>::uniform_runtime_nomemspace_type
+        ViewTypeUniform;
+    Kokkos::Impl::ViewFill<ViewTypeUniform, typename dst_traits::array_layout,
+                           ExecSpace>(dst, value, space);
+  }
+#if defined(KOKKOS_ENABLE_PROFILING)
+  if (Kokkos::Profiling::profileLibraryLoaded()) {
+    Kokkos::Profiling::endDeepCopy();
+  }
+#endif
+}
+
+/** \brief  Deep copy a value from Host memory into a view. ExecSpace can not
+ * access dst */
+template <class ExecSpace, class DT, class... DP>
+inline void deep_copy(
+    const ExecSpace& space, const View<DT, DP...>& dst,
+    typename ViewTraits<DT, DP...>::const_value_type& value,
+    typename std::enable_if<
+        Kokkos::Impl::is_execution_space<ExecSpace>::value &&
+        std::is_same<typename ViewTraits<DT, DP...>::specialize, void>::value &&
+        !Kokkos::Impl::SpaceAccessibility<
+            ExecSpace,
+            typename ViewTraits<DT, DP...>::memory_space>::accessible>::type* =
+        nullptr) {
+  typedef ViewTraits<DT, DP...> dst_traits;
+  static_assert(std::is_same<typename dst_traits::non_const_value_type,
+                             typename dst_traits::value_type>::value,
+                "deep_copy requires non-const type");
+#if defined(KOKKOS_ENABLE_PROFILING)
+  typedef typename dst_traits::memory_space dst_memory_space;
+  if (Kokkos::Profiling::profileLibraryLoaded()) {
+    Kokkos::Profiling::beginDeepCopy(
+        Kokkos::Profiling::SpaceHandle(dst_memory_space::name()), dst.label(),
+        dst.data(), Kokkos::Profiling::SpaceHandle(Kokkos::HostSpace::name()),
+        "(none)", &value, dst.span() * sizeof(typename dst_traits::value_type));
+  }
+#endif
+  if (dst.data() == nullptr) {
+    space.fence();
+  } else {
+    space.fence();
+    typedef typename View<DT, DP...>::uniform_runtime_nomemspace_type
+        ViewTypeUniform;
+    using fill_exec_space = typename dst_traits::memory_space::execution_space;
+    Kokkos::Impl::ViewFill<ViewTypeUniform, typename dst_traits::array_layout,
```

Comment:
It looks like this is the only function call (apart from the surrounding fences) that is different from the other overload.
Can't you just combine the two avoiding the code duplication? I would assume that there is no overhead if the compiler is able to detect at compile time which branch to take.

Reply:
Probably. The reason I didn't do it is that in my taste its easier to think about it this way. Essentially "here is the deep copy if the sync requirement is in line with how I can execute the fill, and here is it if not" from there I can think simply through all the requirement in each case. Otherwise you got to think about conditionals on numerous lines of code of this. I mean we can do that, its a question what is more likely to lead to mistakes: missing a necessary change in a largely duplicated function, or screwing up the logic when adding three conditionals.